### ISSUE TITLE
allow experimental override of `default_request_timeout`

### ISSUE THREAD

**Issue by cfm** [freedomofpress/securedrop-client#1547]
## Description

In development, it's easy to patch [`sdclientapi.API.default_request_timeout`](https://github.com/freedomofpress/securedrop-client/blob/9446c77d59bf236d495ef68195d153837dc99ba2/securedrop_client/api_jobs/sync.py#L43) so that batch metadata operations have more time to complete.  In staging and pilot environments, this is more cumbersome.  We should implement a more-ergonomic override for this value (possibly among others).

## How will this impact [SecureDrop users](https://github.com/freedomofpress/securedrop-ux/wiki/Users)?
Pilot participants would have the ability to tweak this setting in order to experiment with how the SecureDrop Client behaves under different workloads and network conditions.

## How would this affect the SecureDrop Workstation [threat model](https://github.com/freedomofpress/securedrop-workstation/#threat-model)?
Any implementation of this feature already assumes control of both `sd-app` and `dom0`.

## User Stories
1. As a pilot participant, I want to be able to tweak the behavior of the SecureDrop Client's sync operations so that they perform better for my workload and network conditions.

2. As a pilot participant, I want to be able to override approved configuration settings persistently, without manually patching the code.

**Comment by cfm** [comment]
I was initially inclined to approach this via environment variables, e.g.:

![](https://mermaid.ink/img/pako:eNptkEFLAzEQhf9KmHOXeg7Yi12PbmnxUgIym8xqYJNJJ0lVSv-70VoE9fbe8N4H805g2RFoyHSoFC2tPT4LBhNNdBxuVLdaqU-h1aGOlN3YvYovpJaZbBVywmk5MasRxcTsOkzpn44Qut8VEx-4gfhIoi5FregtsRS16-8et_16O2ye7ofh9ov9N_yD6-zsKRZYQCAJ6F176GSiUgbKCwUyoJt0NGGdiwETzy1ak8NCvfOFBXSRSgvAWnj3Hu3VXzLfm1yPCeOeudkJ50znD8PjcZ4)

However, since neither QubesDB nor XenStore values for a domain X are persistent after X is shut down (nor writeable before X is running), we'd still have to find a place to persist (basically, patch...) the `qubesdb-write` calls.

Now I'm inclined to get this value the same way we read [`journalist_key_fingerprint` from `sd-app`'s minimal `config.json`](https://github.com/freedomofpress/securedrop-client/search?q=%22journalist_key_fingerprint%22).  That will be more scalable for these kinds of overrides, especially if we consider introducing a `config.local.json` that won't be reset on every `sdw-admin --apply`.

@eaon, I know you're interested in an even more minimal injection pattern using [Qubes services](https://www.qubes-os.org/doc/qubes-service/).  My understanding is that that would limit us to very simple overrides, such as setting a service `/var/run/qubes-service/SD_TIMEOUT_600` to be interpreted like `SD_TIMEOUT=600` seconds.  Am I right that this would be difficult to use for more-complex values? or do you see this approach being extensible and/or preferable for other reasons?

**Comment by eloquence** [comment]
If I read the code correctly, there are two stages where timeouts can kick in:
- at the `securedrop-sdk` level, with the SDK killing the `qrexec` process regardless of whether we're actively receiving data or not: https://github.com/freedomofpress/securedrop-sdk/blob/main/sdclientapi/__init__.py#L63
- at the `securedrop-proxy` level, with the `requests` library aborting the request if it doesn't receive a response: https://github.com/freedomofpress/securedrop-proxy/blob/main/securedrop_proxy/proxy.py#L240

What actually happens if we make the proxy timeout infinite? What error cases are no longer handled correctly? If we cannot do that because of certain error cases, I wonder if we can detect whether we're still actively receiving bytes from the process, as opposed to just killing the process after the timeout expires.

I think it may be worth doing a bit more investigation into these kinds of strategies before adding a flag, since we'd have to go through the release mechanics for this one way or another, and the time spent on those may well exceed the time spent on the implementation.

**Comment by cfm** [comment]
Discussed in today's Workstation hangout:  I'm going to load my hardware instance with the right quantities of sources, submissions, and replies to give us a common testing endpoint for replicating the current timeout behavior.

**Comment by cfm** [comment]
<https://github.com/freedomofpress/securedrop-client/issues/1547#issuecomment-1223292869>:

> Discussed in today's Workstation hangout: I'm going to load my hardware instance with the right quantities of sources, submissions, and replies to give us a common testing endpoint for replicating the current timeout behavior.

This is done.  See <https://wiki.freedom.press/wiki/User:Cory/SecureDrop_instance> for details, and feel free to contact me for testing access.

**Comment by eaon** [comment]
@cfm thanks for summoning me 🫶 I think your proposed solution of using a JSON key override in `sd-app` is perfectly viable, my main worry there would be that it may normalise touching `sd-app` or our templates for anything but development and setting a precedent for future "quick fixes" along those lines. But since the user base is still limited, and finding our preferred solution for issues like this more elegantly/consistently is a discussion already on the horizon, I'm not too worried about either.

To elaborate a bit on why my preference (in the short term) is still using the service infrastructure to communicate a small amount of information to the client (or any other app) in general:

- It's a persistent system controlled by dom0 that already works right now
- It can be configured using the Qube Manager (could be useful when users encounter this as the solution may be applicable by them without needing a terminal)
- No need to hand-write JSON (in general: argh)
- No risk of accidentally futzing with the system in such a way that it requires a re-run of `sdw-admin --apply` (low risk but still)

While thinking a bit more in depth about this, I remembered that `qvm-service` framework is built on top of [qubes-features](https://dev.qubes-os.org/projects/core-admin/en/latest/qubes-features.html) which, as far as I understand, is Qubes' generic backend for handling key-value information per qube (see [qubes-core-admin/qubes/ext/services.py](https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/ext/services.py) for the services implementation). Whether we can use this framework for our purposes by extending it or by using it via [`qvm-feature`](https://dev.qubes-os.org/projects/core-admin-client/en/latest/manpages/qvm-features.html) or just its generic Python backend, I haven't dug into too much, but I think that's where we should be looking for for various future needs along those lines.

So, in summary, my preferences, in order, would be:

1. Use `qubes.features`/`qvm-features` for our own key-value needs (long term question then would be if we want a GUI for our use as well) (This is probably only feasible in the long term rather than suitable for this ticket)
2. Use the existing service infrastructure as a makeshift key/value store (just because there's already a GUI and it requires less code upfront)
3. Allowing for configuration changes in an AppVM (like `config.local.json`)
4. Any solution that would require running `sdw-admin --apply`

Also I'd totally be game to :pear: pair on this :slightly_smiling_face:

**Comment by cfm** [comment]
With the [default `api_client.default_request_timeout = 60`](https://github.com/freedomofpress/securedrop-client/blob/9446c77d59bf236d495ef68195d153837dc99ba2/securedrop_client/api_jobs/sync.py#L43):

1. On my first invocation of `securedrop-client`, the initial `MetadataSyncJob` succeeded,  after which I saw lots of #1476 even amid a steady stream of `Keyring access from domain: sd-app` notifications indicating that conversation entries were being downloaded and decrypted.
2. On second invocation of `securedrop-client`, the initial `MetadataSyncJob` timed out with:

       securedrop_client.logic:728(on_sync_failure): WARNING: sync failure: The request timed out.

With `api_client.default_request_timeout = 600`, I'm seeing an interesting combination of (transcribed manually from `sd-log`):

```
==> QubesIncomingLogs/sd-proxy/syslog.log
Aug 23 16:11:51 localhost securedrop.Proxy+-sd-app: Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
Aug 23 16:11:51 localhost securedrop.Proxy+-sd-app: BrokenPipeError: [Errno 32] Broken pipe

==> QubesIncomingLogs/sd-app/syslog.log
[...]
Aug 23 16:12:52 localhost 2022-08-23 6:12:52,867 - securedrop_client.logic:728(on_sync_failure) WARNING: sync failure: Unterminated string starting at: line 1 column 4004715 (char 4004714)
```

—and subsequently the more-conventional—

```
securedrop_client.logic:728(on_sync_failure): WARNING: sync failure: The request timed out.
```

After quite some time, a `MetadataSyncJob` finally succeeds, and downloads resume.

More testing tomorrow—and, @eaon, a reply to your <https://github.com/freedomofpress/securedrop-client/issues/1547#issuecomment-1224798524>, for which many thanks!

**Comment by eloquence** [comment]
Without timeout modifications I saw similar results to what you were seeing, Cory -- a few MetadataSyncJobs completed successfully, and messages started downloading, then it eventually started stalling out.

I've made a narrower change to set the `qrexec` timeout to `None` in https://github.com/freedomofpress/securedrop-sdk/blob/main/sdclientapi/__init__.py#L63 and it's been so far syncing without errors, will see if it completes downloading everything. (The hypothesis is that if we stop killing `qrexec` processes that are receiving data, we can make a good number of timing related issues disappear, because `requests` itself should not time-out a request that's receiving data.)

**Comment by eloquence** [comment]
To recap, I set the `qrexec` timeout to `None` in https://github.com/freedomofpress/securedrop-sdk/blob/main/sdclientapi/__init__.py#L63 and let my SecureDrop Client try to sync overnight.

It got halfway done and then stalled out on a network error. I don't think it was related to the sync duration specifically as I also saw individual message/reply download jobs timing out (those shouldn't be unusually large and have a 20 second timeout). I've observed similar behavior in normal use, where the SD Client sometimes is unable to recover from Tor connectivity issues.

After logging out and back in today, I was able to finish syncing. I am seeing the same occasional "Unterminated string" errors in the logs you're seeing, @cfm, which I think we should investigate and track separately.

I would propose that we move forward with an experimental flag, configured in the manner you deem most maintainable. It should be labeled explicitly as experimental, i.e. we cannot guarantee that we will support it in feature. This flag could either do one of the following two things:
- Raise the default timeout to 10 minutes for sync operations
- Eliminate the `qrexec` timeout

As I understand it, if we raise the default timeout to sync operations for 10 minutes, this means that recovering from an unreachable network could take longer, because paused queues are only resumed on a successful sync:

https://github.com/freedomofpress/securedrop-client/blob/a2899c74b03c8d8bfd3f5e7afe4f98d6d23fcf56/securedrop_client/logic.py#L668

In contrast, eliminating the `qrexec` method could have other unexpected side effects -- if the `securedrop-proxy` process hangs indefinitely for any reason, that would be passed along to the SecureDrop Client.

With this in mind, I would suggest you decide on the timeout tweak that makes the most sense to you for this experimental flag, and prepare a change to that effect. We can then discuss with Allie next week on how to potentially proceed with a point release that ships that change. If that makes sense, let me know if you'd like me to test anything else on my patched setup to make that determination.

**Comment by cfm** [comment]
Today @eloquence and I agreed that next week we'll

> move forward with an experimental flag [like an `SDEXTENDEDTIMEOUT_N` service flag]. It should be labeled explicitly as experimental, i.e. we cannot guarantee that we will support it in feature. This flag could either do one of the following two things:
> 
>  * Raise the default timeout to 10 minutes (or a configurable value [`N` seconds]) for sync operations
